code/core/src/phone/java/com/adobe/marketing/mobile/services/ui/notification/PushTemplateHelpers.kt

```diff
@@ -0,0 +1,313 @@
+  the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+  OF ANY KIND, either express or implied. See the License for the specific language
+  governing permissions and limitations under the License.
+*/
+
+package com.adobe.marketing.mobile.services.ui.notification
+
+import android.content.ContentResolver
+import android.content.Context
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.graphics.Matrix
+import android.graphics.RectF
+import android.net.Uri
+import androidx.core.app.NotificationCompat
+import com.adobe.marketing.mobile.services.Log
+import com.adobe.marketing.mobile.services.ServiceProvider
+import com.adobe.marketing.mobile.services.caching.CacheEntry
+import com.adobe.marketing.mobile.services.caching.CacheExpiry
+import com.adobe.marketing.mobile.services.caching.CacheService
+import com.adobe.marketing.mobile.util.UrlUtils
+import java.io.ByteArrayInputStream
+import java.io.ByteArrayOutputStream
+import java.io.File
+import java.io.IOException
+import java.io.InputStream
+import java.net.HttpURLConnection
+import java.net.URL
+import java.util.concurrent.Callable
+import java.util.concurrent.ExecutorService
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+
+/**
+ * General utility functions to assist in building push template notifications.
+ */
+private const val SELF_TAG = "PushTemplateHelpers"
+private const val FULL_BITMAP_QUALITY = 100
+private const val DOWNLOAD_TIMEOUT = 10
+private const val CLASS_CREATION_TIMEOUT = 2
+
+/**
+ * Asset cache location to use for downloaded push template images.
+ *
+ * @return [String] containing the asset cache location to use for storing downloaded push template images.
+ */
+internal val assetCacheLocation: String?
+    get() {
+        val deviceInfoService = ServiceProvider.getInstance().deviceInfoService
+            ?: return null
+        val applicationCacheDir = deviceInfoService.applicationCacheDir
+        return if ((applicationCacheDir == null)) null else (
+            (
+                applicationCacheDir
+                    .toString() + File.separator +
+                    PushTemplateConstants.CACHE_BASE_DIR
+                ) + File.separator +
+                PushTemplateConstants.PUSH_IMAGE_CACHE
+            )
+    }
+
+private val executor: ExecutorService
+    get() = ExecutorHolder.INSTANCE
+
+private object ExecutorHolder {
+    val INSTANCE: ExecutorService = Executors.newSingleThreadExecutor()
+}
+
+private class DownloadImageCallable(val url: String?) :
+    Callable<Bitmap?> {
+    override fun call(): Bitmap? {
+        val bitmap: Bitmap?
+        var connection: HttpURLConnection? = null
+        var inputStream: InputStream? = null
+        try {
+            val imageUrl = URL(url)
+            connection = imageUrl.openConnection() as HttpURLConnection
+            inputStream = connection.inputStream
```

Comment:
A few recommendations
-  use the `use` function to ensure that the connection stream is auto closed. That way we can get rid of the connection close handling in the catch block.
-  use String templates for logging as it is more idiomatic Kotlin
- log the success case only if the bitmap was downloaded successfully.
- disconnect the connection once we are done.

```
        var bitmap: Bitmap? = null
        var connection: HttpURLConnection? = null
        try {
            val imageUrl = URL(url)
            connection = imageUrl.openConnection() as? HttpURLConnection
            connection?.inputStream.use { inputStream ->
                bitmap = BitmapFactory.decodeStream(inputStream)
                Log.trace(
                    PushTemplateConstants.LOG_TAG,
                    SELF_TAG,
                    "Downloaded push notification image from url ($url)"
                )
            }
        } catch (e: IOException) {
            Log.warning(
                PushTemplateConstants.LOG_TAG,
                SELF_TAG,
                "Failed to download push notification image from url ($url). Exception: ${e.message}"
            )
        } finally {
            connection?.disconnect()
        }
        return bitmap
    }
```